File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.py
import sys
input = sys.stdin.readline
from collections import deque

n, m = map(int, input().split())
board = [list(input().rstrip()) for _ in range(n)]
dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def move_to_end(i, x, y):
    while True:
        nx, ny = x + dx[i], y + dy[i]
        # 다음 위치가 벽이면 현재 위치 리턴
        if board[nx][ny] == '#':
            return x, y
        # 다음 위치가 구멍이면 구멍 위치 리턴
        if board[nx][ny] == 'O':
            return nx, ny
        x, y = nx, ny

def bfs(red, blue):
    queue = deque([(red, blue)])
    visited = [[[[-1]*m for _ in range(n)] for _ in range(m)] for _ in range(n)]
    visited[red[0]][red[1]][blue[0]][blue[1]] = 0
    while queue:
        (rx, ry), (bx, by) = queue.popleft()
        if visited[rx][ry][bx][by] > 10:
            return -1
        if board[rx][ry] == 'O':
            return visited[rx][ry][bx][by]
        for i in range(4):
            nrx, nry = move_to_end(i, rx, ry)
            nbx, nby = move_to_end(i, bx, by)
            
            # 파란 공이 구멍에 빠지면 불가능
            if board[nbx][nby] == 'O':
                continue
            # 빨간 공, 파란 공 위치 같은 경우 처리
            if nrx == nbx and nry == nby:
                r_size = rx*dx[i] + ry*dy[i]
                b_size = bx*dx[i] + by*dy[i]
                # 빨간 공은 그대로, 파란 공은 한 칸 덜 가기
                if r_size > b_size:
                    nbx, nby = nbx - dx[i], nby - dy[i]
                else:
                    nrx, nry = nrx - dx[i], nry - dy[i]
            # 이전에 방문했던 곳이면 불가능
            if visited[nrx][nry][nbx][nby] != -1:
                continue
            visited[nrx][nry][nbx][nby] = visited[rx][ry][bx][by] + 1
            queue.append(((nrx, nry), (nbx, nby)))
    return -1

red = []
blue = []
for i in range(n):
    for j in range(m):
        if board[i][j] == 'R':
            red = (i, j) 
        elif board[i][j] == 'B':
            blue = (i, j)
print(bfs(red, blue))

File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n");

let index = 0;
const [n, m] = input[index++].split(" ").map(Number);
const board = [];
for (let i = 0; i < n; i++) {
  board.push(input[index++].split(""));
}

let red;
let blue;

for (let i = 0; i < n; i++) {
  for (let j = 0; j < m; j++) {
    if (board[i][j] === "R") {
      red = { x: i, y: j };
    } else if (board[i][j] === "B") {
      blue = { x: i, y: j };
    }
  }
}
const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];

console.log(bfs());

function bfs() {
  const visited = new Set();
  const queue = [[red, blue, 0]];
  visited.add(`${red.x},${red.y},${blue.x},${blue.y}`);

  while (queue.length) {
    const [cur_red, cur_blue, move] = queue.shift();

    if (10 < move) {
      return -1;
    }

    if (board[cur_red.x][cur_red.y] === "O") {
      return move;
    }

    red = cur_red;
    blue = cur_blue;

    for (let i = 0; i < 4; i++) {
      let next_red = move_ball(i, cur_red.x, cur_red.y);
      let next_blue = move_ball(i, cur_blue.x, cur_blue.y);

      if (board[next_blue.x][next_blue.y] === "O") continue;

      if (next_red.x === next_blue.x && next_red.y === next_blue.y) {
        [next_red, next_blue] = adjust_balls(i, next_red.x, next_red.y);
      }

      const key = `${next_red.x},${next_red.y},${next_blue.x},${next_blue.y}`;
      if (!visited.has(key)) {
        visited.add(key);
        queue.push([next_red, next_blue, move + 1]);
      }
    }
  }

  return -1;
}

function adjust_balls(direction, x, y) {
  if (direction === 0) {
    if (red.y < blue.y) {
      return [
        { x, y },
        { x, y: y + 1 },
      ];
    } else {
      return [
        { x, y: y + 1 },
        { x, y },
      ];
    }
  } else if (direction === 1) {
    if (red.y > blue.y) {
      return [
        { x, y },
        { x, y: y - 1 },
      ];
    } else {
      return [
        { x, y: y - 1 },
        { x, y },
      ];
    }
  } else if (direction === 2) {
    if (red.x < blue.x) {
      return [
        { x, y },
        { x: x + 1, y },
      ];
    } else {
      return [
        { x: x + 1, y },
        { x, y },
      ];
    }
  } else if (direction === 3) {
    if (red.x > blue.x) {
      return [
        { x, y },
        { x: x - 1, y },
      ];
    } else {
      return [
        { x: x - 1, y },
        { x, y },
      ];
    }
  }
}

function move_ball(d, x, y) {
  while (true) {
    if (board[x + dx[d]][y + dy[d]] === "#") break;
    if (board[x][y] === "O") break;

    x += dx[d];
    y += dy[d];
  }

  return { x, y };
}
